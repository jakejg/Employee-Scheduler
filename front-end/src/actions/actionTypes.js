export const LOAD_JOBS = "LOAD_JOBS";
export const ADD_JOBS = "ADD_JOB";