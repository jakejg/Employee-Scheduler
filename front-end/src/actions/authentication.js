import axios from 'axios';
import { 
    ADD_OR_REMOVE_TOKEN
 } from './actionTypes';
import {addCompanyToAPI} from '../helpers/JobApi'
import {BASE_URL} from '../config';


export const register = (data) => {
    
    return async (dispatch) => {
        try {
            // designate admin user
            data.is_admin = true
            // create admin user
       
            const res = await axios.post(`${BASE_URL}/auth/register`, data);
       
            dispatch(addOrRemoveToken(res.data.token));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const login = (data) => {
    return async (dispatch) => {
        try {
            let res = await axios.post(`${BASE_URL}/auth/login`, data);
            dispatch(addOrRemoveToken(res.data.token));
        }
        catch(e) {
            // console.log(e.response.data.message)
            return e.response.data.message
        }
    }
}

export const addOrRemoveToken = (token) => {
    return {type: ADD_OR_REMOVE_TOKEN, token}
}
