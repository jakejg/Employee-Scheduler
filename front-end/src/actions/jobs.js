import { 
    LOAD_JOBS,
    EDIT_JOB, 
    ADD_JOB,
    EDIT_JOB_STAFF,
    DELETE_JOB
 } from './actionTypes';
import {JobAPI} from '../helpers/JobApi';

export const loadJobsFromAPI = (comp_id) => {
    return async (dispatch) => {
        try {
            const jobData = await JobAPI.loadJobs(comp_id)
            dispatch(loadJobs(jobData));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const loadJobs = (jobs) => {
    return {type: LOAD_JOBS, jobs}
}

export const addJobOnAPI = (jobToAdd) => {

    return async (dispatch) => {
        try {
            let job = await JobAPI.addJob(jobToAdd)
            dispatch(addJob(job.id, job))
            return {message: "Job created Successfully", severity: "success"}
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const addJob = (id, job) => {
    return {type: ADD_JOB, id, job}
}


export const getJobFromAPI = (ID) => {
    return async (dispatch) => {
        try {
            let job = await JobAPI.getJob(ID)
            dispatch(editJob(job.id, job))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const editJobOnAPI = (id, changedProps) => {
    return async (dispatch) => {
        try {
            let job = await JobAPI.editJob(id, changedProps)
            console.log(job)
            dispatch(editJob(job.id, job))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const editJob = (id, job) => {
    return {type: EDIT_JOB, id, job}
}

export const addStaffToJobOnAPI = (jobId, staffId) => {
    return async (dispatch) => {
        try {
            let data = await JobAPI.addStaffToJob(jobId, staffId)
            console.log(data)
            dispatch(editJobStaff(jobId, data.staffList, data.staff_filled))
            
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const removeStaffFromJobOnAPI = (jobId, staffId) => {
    return async (dispatch) => {
        try {
            let data = await JobAPI.removeStaffFromJob(jobId, staffId)
            dispatch(editJobStaff(jobId, data.staffList, data.staff_filled))
            
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const editJobStaff = (id, staff, staff_filled) => {
    return {type: EDIT_JOB_STAFF, id, staff, staff_filled}
}

export const deleteJobOnAPI = (id) => {
    return async (dispatch) => {
        try {
            let msg = await JobAPI.deleteJob(id)
            dispatch(deleteJob(id))
        }
        catch(e) {
            console.log(e)
            return e.message
        }
    }
}

export const deleteJob = (id) => {
    return {type: DELETE_JOB, id}
}