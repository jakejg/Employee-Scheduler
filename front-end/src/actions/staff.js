import { 
    LOAD_STAFF,
    EDIT_STAFF,
    ADD_STAFF
 } from './actionTypes';
import { StaffAPI } from '../helpers/StaffApi';

export const loadStaffFromAPI = (comp_id) => {
    return async (dispatch) => {
        try {
            let staffData = await StaffAPI.loadStaff(comp_id);
            dispatch(loadStaff(staffData));
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const loadStaff = (staff) => {
    return {type: LOAD_STAFF, staff}
}


export const getStaffFromAPI = (ID) => {
    return async (dispatch) => {
        try {
            let staff = await StaffAPI.getStaff(ID)
        
            dispatch(editStaff(staff.id, staff))
            }
        catch(e) {
            console.log(e)
        }
    }
}

export const editStaff = (id, staff) => {
    return {type: EDIT_STAFF, id, staff}
}

export const addStaffOnAPI = (staffToAdd) => {
   
    return async (dispatch) => {
        try {
            const staff = await StaffAPI.addStaff(staffToAdd) ;
            dispatch(addStaff(staff.id, staff))
        }
        catch(e) {
            console.log(e)
        }
    }
}

export const addStaff = (id, staff) => {
    return {type: ADD_STAFF, id, staff}
}