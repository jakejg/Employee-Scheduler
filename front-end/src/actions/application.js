import {CHANGE_DRAWER, CHANGE_NO_JOBS, CHANGE_NO_STAFF} from './actionTypes';

export const changeDrawer = () => {
    return {type: CHANGE_DRAWER}
}

export const changeNoJobs = () => {
    return {type: CHANGE_NO_JOBS}
}

export const changeNoStaff = () => {
    return {type: CHANGE_NO_STAFF}
}