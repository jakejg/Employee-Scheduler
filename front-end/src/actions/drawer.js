import {CHANGE_DRAWER} from './actionTypes';

export const changeDrawer = () => {
    return {type: CHANGE_DRAWER}
}