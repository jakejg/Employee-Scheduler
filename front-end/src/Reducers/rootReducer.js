import jobs from './jobsReducer';
import staff from './staffReducer';
import { combineReducers } from "redux";

export default combineReducers({jobs, staff})