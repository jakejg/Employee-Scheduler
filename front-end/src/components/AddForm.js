import React, { useState} from 'react';
import {useSelector, useDispatch} from 'react-redux';
import { changeNoJobs } from '../actions/application';
import { changeNoStaff } from '../actions/application';
import {
    Grid,
    TextField,
    Box,
    Paper,
    Button
} from '@material-ui/core';
import Alert from '@material-ui/lab/Alert';
import {useHistory} from 'react-router-dom';
import underscoreName from '../helpers/underscoreName';
import ButtonGroup from './ButtonGroup';


const AddForm = ({  type,
                    fields,
                    doOnSubmit,
                    redirect,
                    closeDialog
                }) => {

    // set up form state and error state from fields passed in
    const INITIAL_FORM_STATE = {};
    const INITIAL_FORM_FIELDS_STATE = [];
    for (let field of fields){ 
        INITIAL_FORM_STATE[underscoreName(field)] = ""; 
        INITIAL_FORM_FIELDS_STATE.push({name: field, error: false})
    }

    
    const [formData, setFormData] = useState(INITIAL_FORM_STATE);
    const [formFields, setFormFields] = useState(INITIAL_FORM_FIELDS_STATE);
    const [alert, setAlert] = useState();
    const dispatch = useDispatch()
    const history = useHistory();
    const noJobs = useSelector(state => state.application.noJobs);
    const noStaff = useSelector(state => state.application.noStaff);

    const handleChange = (e) => {
        const { name, value } = e.target;
        setFormData(formData => ({...formData, [name]: value}))
    }
    
    const handleSubmit = async (e) => {
        e.preventDefault();
        if (!validate()){
            let msg = await dispatch(doOnSubmit(formData));
            if (msg){
                setAlert(alert => msg)
            }
            if (noJobs && type === 'Job'){
                dispatch(changeNoJobs())
            }
            if (noStaff && type === 'Staff'){
                dispatch(changeNoStaff())
            }
            else{
                
                if (redirect) {
                    closeDialog();
                    history.push(redirect);
                }
            }
        }
        
    }

    // check that every field is filled out
    const validate = () => {
        let foundErrors = false;
        const updated = formFields.map(field => {
            
            if (!formData[underscoreName(field.name)]) {
                field.error = "This field is required";
                foundErrors = true;
            }
            // remove errors
            else {
                field.error = false;
            }
            return field
        });
        if (foundErrors) {
            setFormFields(updated);
        }
        return foundErrors;
    }

    // mx={{sm:'auto'}} width={{sm:'40%'}} mt={3}
    return (
        <form onSubmit={handleSubmit} >
            <Box  >
            <Paper>
            <Grid container >
                <Box  mx='auto' m={1}>
                {alert && <Alert severity={alert.severity}>{alert.message}</Alert>}
                {formFields.map(field => {
                                    let type = "text";
                                    if (field.name.endsWith('Date')) type = "date";
                                    if (field.name === "Password") type = "password";
                                    if (field.name.endsWith('Wage')) type = "number";
                                    if (field.name.endsWith('Company')) type = "number";
                                    return (<Grid item key={field.name}> 
                                        <TextField 
                                                    error={field.error ? true: false}
                                                    id={field.name}
                                                    type={type}
                                                    label={field.name}
                                                    // change name format for sending to backend
                                                    name={underscoreName(field.name)}
                                                    value={formData.underscoreName}
                                                    onChange={handleChange}
                                                    variant="outlined"
                                                    margin="normal"
                                                    helperText={field.error}
                                                    InputLabelProps={{
                                                        shrink: true,
                                                      }}
                                                    />
                                    </Grid>)}
                )}
                {type === 'Job' || type === 'Staff' ?
                <ButtonGroup>
                    <Button variant="contained" color={type === 'job' ? 'primary' : 'secondary'} type="submit"> Add {type}</Button>
                    <Button variant="contained" style={{backgroundColor:'#f0ad4e'}} onClick={closeDialog}>Cancel</Button>
                </ButtonGroup>   
                :
                <Button variant="contained" type="submit">{type}</Button>
                }
                </Box>
            </Grid> 
            </Paper> 
            </Box>
           
        </form>
    )
}

export default AddForm;