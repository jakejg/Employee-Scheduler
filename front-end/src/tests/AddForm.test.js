import React from 'react';
import { render, fireEvent, act, within } from '@testing-library/react';
import {Provider} from 'react-redux'
import AddForm from '../components/AddForm';
import {createStore, applyMiddleware} from "redux";
import rootReducer from "../reducers/rootReducer";
import { MemoryRouter } from "react-router";
import { addStaffOnAPI } from '../actions/staff';
import { addOrRemoveToken } from '../actions/authentication';
import thunk from 'redux-thunk'
import moment from 'moment';

const store = createStore( rootReducer, applyMiddleware(thunk));
store.dispatch(addOrRemoveToken("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6IkRlbW8iLCJpc19hZG1pbiI6dHJ1ZSwiY29tcF9pZCI6MSwiaWF0IjoxNTk3NzY5NTYyfQ.Wm2z3GCj-m6DDWElCIBCuRxqBbc3_0QhJzeorsbz6E0")); 

const jobFields = ['Title', 'Start Date','End Date', 'Staff Needed', 'Notes'];
const staffFields = ['Email', 'First Name','Last Name', 'Current Wage', 'Years At Company'];

it('renders without crashing', () => {
   
  render(
    <MemoryRouter>
      <Provider store={store}>
      <AddForm
            type='Staff'
            fields={staffFields}
            doOnSubmit={addStaffOnAPI} />
      </Provider>
    </MemoryRouter>
)
});

it('matches snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <Provider store={store}>
                <AddForm
                    type='Staff'
                    fields={staffFields}
                    doOnSubmit={addStaffOnAPI} />
            </Provider>
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot();
})





it('it should add a staff to the DOM', async () => {
    const { getByLabelText, getByText} = render(<Provider store={store}>
        <AddForm
            type='Staff'
            fields={staffFields}
            doOnSubmit={addStaffOnAPI} />
    </Provider>);
    const emailInput = getByLabelText("Email");
    console.log(emailInput);
    const firstNameInput = getByLabelText("First Name");
    const lastNameInput = getByLabelText("Last Name");
    const wageInput = getByLabelText("Current Wage");
    const yearsInput = getByLabelText("Years At Company");
    const btn = getByText("Add Staff");
    await act(async () => {
    fireEvent.change(emailInput, {target: {value: "test@t.com"}});
    console.log(emailInput)
    fireEvent.change(firstNameInput, {target: {value: "Nick"}});
    fireEvent.change(lastNameInput, {target: {value: "Jones"}});
    fireEvent.change(wageInput, {target: {value: 140}});
    fireEvent.change(yearsInput, {target: {value: 3}});
    fireEvent.submit(btn);  
      });


  
    expect(getByText("Staff Added Successfully")).toBeInTheDocument();
  
  })