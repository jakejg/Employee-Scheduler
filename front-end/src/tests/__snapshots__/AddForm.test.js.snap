// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiBox-root MuiBox-root-9"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiBox-root MuiBox-root-10"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined"
                  data-shrink="true"
                  for="Email"
                  id="Email-label"
                >
                  Email
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    id="Email"
                    name="email"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-13 PrivateNotchedOutline-legendNotched-14"
                    >
                      <span>
                        Email
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined"
                  data-shrink="true"
                  for="First Name"
                  id="First Name-label"
                >
                  First Name
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    id="First Name"
                    name="first_name"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-13 PrivateNotchedOutline-legendNotched-14"
                    >
                      <span>
                        First Name
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined"
                  data-shrink="true"
                  for="Last Name"
                  id="Last Name-label"
                >
                  Last Name
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    id="Last Name"
                    name="last_name"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-13 PrivateNotchedOutline-legendNotched-14"
                    >
                      <span>
                        Last Name
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined"
                  data-shrink="true"
                  for="Current Wage"
                  id="Current Wage-label"
                >
                  Current Wage
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    id="Current Wage"
                    name="current_wage"
                    type="number"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-13 PrivateNotchedOutline-legendNotched-14"
                    >
                      <span>
                        Current Wage
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined"
                  data-shrink="true"
                  for="Years At Company"
                  id="Years At Company-label"
                >
                  Years At Company
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    id="Years At Company"
                    name="years_at_company"
                    type="number"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-13 PrivateNotchedOutline-legendNotched-14"
                    >
                      <span>
                        Years At Company
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-16 makeStyles-box-15"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                   Add Staff
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained"
                style="background-color: rgb(240, 173, 78);"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Cancel
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
